<!-- -->
<launch>

    <!-- Set localization method: -->
    <arg name="topic_pose_stamped" default="/ndt_pose"/> <!-- ndt_pose | icp_pose | gnss_pose -->

    <!-- Set args defining tf_transform: base_link to velodyne-->
    <arg name="tf_x" default="0.6"/>
    <arg name="tf_y" default="0.0"/>
    <arg name="tf_z" default="2.0"/>
    <arg name="tf_yaw" default="0.0"/>
    <arg name="tf_pitch" default="0.0"/>
    <arg name="tf_roll" default="0.0"/>

    <!-- tf_transform: base_link to velodyne -->
    <include file="$(find runtime_manager)/launch_files/setup_tf.launch">
        <arg name="x" value="$(arg tf_x)"/>
        <arg name="y" value="$(arg tf_y)"/>
        <arg name="z" value="$(arg tf_z)"/>
        <arg name="yaw" value="$(arg tf_yaw)"/>
        <arg name="pitch" value="$(arg tf_pitch)"/>
        <arg name="roll" value="$(arg tf_roll)"/>
        <arg name="frame_id" value="base_link"/>
        <arg name="child_frame_id" value="/velodyne"/>
        <arg name="period_in_ms" value="10"/>
    </include>

    <!-- Set params-->
    <param name="localizer" value="velodyne"/>
    <param name="tf_x" value="$(arg tf_x)"/>
    <param name="tf_y" value="$(arg tf_y)"/>
    <param name="tf_z" value="$(arg tf_z)"/>
    <param name="tf_yaw" value="$(arg tf_yaw)"/>
    <param name="tf_pitch" value="$(arg tf_pitch)"/>
    <param name="tf_roll" value="$(arg tf_roll)"/>

    <!-- Vehicle Model -->
    <include file="$(find vehicle_description)/launch/vehicle_model.launch">
        <arg name="model_path" value="$(find adeye)/../../../../AD-EYE/Data/URDF_Actor_Models/BMW_X5/default.urdf"/>
    </include>

    <!-- Points Downsampler -->
    <node pkg="rostopic" type="rostopic" name="config_voxel_grid_filter_rostopic"
          args="pub -l /config/voxel_grid_filter autoware_config_msgs/ConfigVoxelGridFilter
        '{voxel_leaf_size: 0.2 , measurement_range: 200.0}' "
    />
    <include file="$(find points_downsampler)/launch/points_downsample.launch"/>

    <!-- ndt_matching --> <!-- METHOD TIPE: pcl_generic=0, pcl_anh=1, pcl_anh_gpu=2, pcl_openmp=3 -->
    <include file="$(find lidar_localizer)/launch/ndt_matching.launch">
      <arg name="method_type" value="0" />
      <arg name="use_gnss" value="1" />
      <arg name="use_odom" value="false" />
      <arg name="use_imu" value="false" />
      <arg name="imu_upside_down" value="false" />
      <arg name="imu_topic" value="imu_raw" />
      <arg name="queue_size" value="1" />
      <arg name="offset" value="linear" />
      <arg name="get_height" value="true" />
      <arg name="use_local_transform" value="false" />
      <arg name="sync" value="false" />
    </include>

    <!-- Pose Connect -->
    <node pkg="topic_tools" type="relay" name="pose_relay" output="log" args="$(arg topic_pose_stamped) /current_pose"/>

</launch>
